design/general-progress-monitoring.md

```diff
@@ -0,0 +1,499 @@
+`<backup-name>-itemoperations.json.gz` file will be added that contains the items and operation IDs
+(snapshotIDs) returned by VolumeSnapshotter and BackupItemAction plugins.  Also, the creation of the
+`velero-backup.json` object will not occur until the backup moves to one of the terminal phases
+(_Completed_, _PartiallyFailed_, or _Failed_).  Reconciliation should ignore backups that do not
+have a `velero-backup.json` object.
+
+The cluster that is creating the backup will have the Backup resource present and will be able to
+manage the backup before the backup completes.
+
+If the Backup resource is removed (e.g. Velero is uninstalled) before a backup completes and writes
+its `velero-backup.json` object, the other objects in the object store for the backup will be
+effectively orphaned.  This can currently happen but the current window is much smaller.
+
+### `<backup-name>-itemoperations.json.gz`
+The itemoperations file is similar to the existing `<backup-name>-itemsnapshots.json.gz`  Each snapshot taken via
+BackupItemAction will have a JSON record in the file.  Exact format TBD.
+
+## Changes to Velero restores
+
+A `<restore-name>-itemoperations.json.gz` file will be added that contains the items and operation IDs
+returned by RestoreItemActions.
+
+## CSI snapshots
+
+For systems such as EBS, a snapshot is not available until the storage system has transferred the
+snapshot to stable storage.  CSI snapshots expose the _readyToUse_ state that, in the case of EBS,
+indicates that the snapshot has been transferred to durable storage and is ready to be used.  The
+CSI BackupItemAction.Progress method will poll that field and when completed, return completion.
+
+## vSphere plug-in
+
+The vSphere Plug-in for Velero uploads snapshots to S3 in the background.  This is also a
+BackupItemAction plug-in, it will check the status of the Upload records for the snapshot and return
+progress.
+
+## Backup workflow changes
+
+The backup workflow remains the same until we get to the point where the `velero-backup.json` object
+is written.  At this point, we will queue the backup to a finalization go-routine.  The next backup
+may then begin.  The finalization routine will run across all of the
+VolumeSnapshotter/BackupItemAction operations and call the _Progress_ method on each of them.
+
+If all snapshot and backup item operations have finished (either successfully or failed), the backup
+will be completed and the backup will move to the appropriate terminal phase and upload the
+`velero-backup.json` object to the object store and the backup will be complete.
+
+If any of the snapshots or backup items are still being processed, the phase of the backup will be
+set to the appropriate phase (_WaitingForPluginOperations_ or
+_WaitingForPluginOperationsPartialFailure_).  In the event of any of the progress checks return an
+error, the phase will move to _WaitingForPluginOperationsPartialFailure_.  The backup will then be
```

Comment:
Once the phase is moved to _WaitingForPluginOperationsPartialFailure_ and is requeued, will the user know it is requeued? Will the phase move back to _WaitingForPluginOperations_ ?

Reply:
Once it's moved to WaitingForPluginOperationsPartialFailure, it won't move back to WaitingForPluginOperations since any backup or restore with at least one recorded error on it is on the partial failure track. If there are errors *before* moving to one of the waiting phases, then it moves to WaitingForPluginOperationsPartialFailure and eventually to PartiallyFailed. If there are no errors before waiting, then it moves to WaitingForPluginOperations and eventually to Completed, unless an error is logged by one of the async actions we're waiting on -- if that happens it moves to WaitingForPluginOperationsPartialFailure and then once all actions complete, it moves to PartiallyFailed.